packages/x-bus/src/x-bus.ts

```diff
@@ -0,0 +1,245 @@
+import { XPriorityQueue } from '@empathyco/x-priority-queue';
+import { AnyFunction, Dictionary } from '@empathyco/x-utils';
+import { Observable, ReplaySubject } from 'rxjs';
+import { map } from 'rxjs/operators';
+import {
+  EmittedData,
+  Emitter,
+  Emitters,
+  Priority,
+  SubjectPayload,
+  TimeoutId,
+  XBus,
+  XEvent,
+  XEventPayload
+} from './x-bus.types';
+
+declare module '@empathyco/x-priority-queue' {
+  export interface XPriorityQueueNodeMetadata {
+    resolve: (value: EmittedData<XEvent>) => void;
+    eventPayload: any;
+    eventMetadata: any;
+  }
+}
+
+/**
+ * A {@link XBus} implementation using an {@link XPriorityQueue} as its data structure to
+ * prioritise the emission of events. The priorities are preconfigured based on event naming.
+ *
+ * @public
+ */
+export class XPriorityBus implements XBus {
+  /**
+   *
+   *
+   * @internal
+   */
+  protected queue: XPriorityQueue<XEvent>;
+
+  /**
+   *
+   *
+   * @internal
+   */
+  protected priorities: Dictionary<number>;
+
+  /**
+   *
+   *
+   * @internal
+   */
+  protected emitCallbacks: AnyFunction[];
+
+  /**
+   * A dictionary to store the created event emitters.
+   *
+   * @internal
+   */
+  protected emitters: Emitters = {};
+
+  /**
+   * A pending flush operation {@link TimeoutId | timeout identifier} or undefined if there's
+   * none pending.
+   *
+   * @internal
+   */
+  protected pendingFlush?: TimeoutId;
+
+  /**
+   * A list of pending pop operations {@link TimeoutId | timeout identifiers}.
+   *
+   * @internal
+   */
+  protected pendingPops: TimeoutId[] = [];
+
+  /**
+   * Creates a new instance of a {@link XPriorityBus}.
+   *
+   * @param config
+   * @param queue - A {@link XPriorityQueue} to store the events.
+   * @param priorities - A {@link @empathyco/x-utils#Dictionary} defining the priorities associated
+   * to a given string.
+   */
+  public constructor(
+    config: {
+      queue?: XPriorityQueue;
+      priorities?: Dictionary<number>;
```

Comment:
Should we use here the `Priority` type?
```suggestion
      priorities?: Dictionary<Priority>;
```

Reply:
Good eye, done!